firewood/src/proof.rs

```diff
@@ -455,93 +447,51 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
         buf: &[u8],
         end_node: bool,
     ) -> Result<(DiskAddress, Option<SubProof>, usize), ProofError> {
-        let mut items: Vec<Encoded<Vec<u8>>> = bincode::DefaultOptions::new().deserialize(buf)?;
-        let size = items.len();
-
-        match size {
-            EXT_NODE_SIZE => {
-                let mut items = items.into_iter();
-
-                let cur_key_path: Vec<u8> = items
-                    .next()
-                    .unwrap()
-                    .decode()?
-                    .into_iter()
-                    .flat_map(to_nibble_array)
-                    .collect();
-
-                let (cur_key_path, term) = PartialPath::decode(&cur_key_path);
-                let cur_key = cur_key_path.into_inner();
-
-                let data: Vec<u8> = items.next().unwrap().decode()?;
-
+        let node = NodeType::decode(buf)?;
```

Comment:
nit: naming: `decode` -> `new` or, even beter, `impl From<&[u8]> for `NodeType`. Would also be okay with `deserialize`.

Reply:
Probably will do in a follow up.

Reply:
I think the `decode` static method is aptly named. `new` usually does not contain much logic and is generally just used to set fields. `TryFrom` might make more sense, but even native types like `u64` have static methods like [`from_ne_bytes`](https://doc.rust-lang.org/stable/std/primitive.u64.html#method.from_ne_bytes) instead of implementing the `TryFrom` trait. 

Usually `From` and `Into` and their `Try*` equivalents are more like type-casts. A `Vec<u8>` may contain the data to create a `NodeType`, but they are two very different types. 

Checkout [all the `TryFrom` implementors](https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html#implementors) to see the pattern.

Reply:
That's because there isn't a type specific to endianness to convert from, and `TryFrom` doesn't take any arguments -- there has to be a concrete type, so that's appropriate. However, here there aren't any other arguments.